config/user_agent_filter.go

```diff
@@ -0,0 +1,35 @@
+package config
+
+import (
+	"golang.ngrok.com/ngrok/internal/pb"
+)
+
+// UserAgentFilter is a pair of strings slices that allow/deny traffic to an endpoint
+type userAgentFilter struct {
+	// slice of regex strings for allowed user agents
+	Allow []string
+	// slice of regex strings for denied user agents
+	Deny []string
+}
+
+func (b *userAgentFilter) toProtoConfig() *pb.MiddlewareConfiguration_UserAgentFilter {
+	if b == nil {
+		return nil
+	}
+	return &pb.MiddlewareConfiguration_UserAgentFilter{
+		Allow: b.Allow,
+		Deny:  b.Deny,
+	}
+}
+
+// WithUserAgentFilter configures user-agent filter from two lists of regeular expressions that allow/deny traffic to an endpoint.
```

Comment:
typo in `regular` (regeular).

I feel like this also doesn't cover everything you want to know here, such as associated error codes.

May I suggest:

```suggestion
// WithUserAgentFilter adds user agent filtering to the endpoint.
//
// The allow argument is a list of regular expressions for the user-agent
// header to allow, and the deny argument is a list of regular expressions to
// deny, with allows taking precedence over denies.
//
// And invalid regular expression will result in an error when creating the unnel.
//
// TODO(mo): add link to relevant ngrok agent documentation once the documentation is live.
// TODO(mo): add specific errcode, once there is one, for invalid allow/deny on connect.
```